doc/source/rllib/env-to-module-connector.rst

```diff
@@ -0,0 +1,546 @@
+            batch = {}
+            batch = env_to_module(
+                episodes=[episode],
+                batch=batch,
+                rl_module=rl_module,  # in stateful case, RLModule is required
+                explore=True,
+            )
+            # Print out the resulting batch.
+            print(batch)
+
+
+
+You can see that the pipeline extracted the current observations from the two
+running episodes and placed them under the ``obs`` column into the forward batch.
+The batch has a size of 2, because we had 2 episodes, and should look similar to this:
+
+.. code-block:: text
+
+    {'obs': tensor([[ 0.0212, -0.1996, -0.0414,  0.2848],
+            [ 0.0292,  0.0259, -0.0322, -0.0004]])}
+
+In the stateful case, you can also expect the ``STATE_IN`` columns to be present.
+Note that because of the LSTM layer, the internal state of the module consists of two components, ``c`` and ``h``:
+
+.. code-block:: text
+
+    {
+        'obs': tensor(
+            [[ 0.0212, -0.1996, -0.0414,  0.2848],
+            [ 0.0292,  0.0259, -0.0322, -0.0004]]
+        ),
+        'state_in': {
+            # Note: The shape of each state tensor here is
+            # (B=2, [num LSTM-layers=1], [LSTM cell size]).
+            'h': tensor([[[0., 0., .., 0.]]]),
+            'c': tensor([[[0., 0., ... 0.]]]),
+        },
+    }
+
+.. hint::
+
+    You are free to design the internal states of your custom :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` classes
+    however you like. You only need to override the :py:meth:`` method and make sure you return a new state of any nested structure and
+    shape from your `forward_..()` methods under the fixed ``state_out`` key.
+    See `here for an example <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/classes/lstm_containing_rlm.py>`__
+    of an RLModule class with a custom LSTM layer in it.
+
+
+Writing custom env-to-module connectors
+---------------------------------------
+
+You can customize the default env-to-module pipeline that RLlib creates through specifying a function in your
+:py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`, which takes an optional RL environment object (`env`) and an optional `spaces`
+dictionary as input arguments and returns a single :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` piece or a list thereof.
+RLlib prepends the provided :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` instances to the
+:ref:`default env-to-module pipeline <default-env-to-module-pipeline>` in the order returned,
+unless you set `add_default_connectors_to_env_to_module_pipeline=False` in your config, in which case RLlib exclusively uses the provided
+:py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pieces without any automatically added default behavior.
+
+.. _env-to-module-connectors-structure-of-spaces-arg:
+
+Note that RLlib expects the structure of the `spaces` argument to be:
+
```

Comment:
Maybe adding some hints of where to pack certain types of connectors in the pipeline (e.g. observation transformation before `AddObsFromEpisodesToTrainBatch` when the user wants the changes to be permanent and therefore also relevant for the `Learner`?